File: contest/weekly/weekly-contest-218/1679_maxNumOfKSumPairs.go
/**
 *      file    :  1679_maxNumOfKSumPairs.go
 *      author  :  techiedheeraj
 *      created :  2020 Dec 06 17:57:08
 *      lastMod :  Sun Dec  6 17:57:08 2020
**/

func maxOperations(nums []int, k int) int {
  sort.Ints(nums)

  var ans int = 0
  var i int = 0
  var j int = len(nums)-1

  for i < j {
    var val int = nums[i] + nums[j]
    if val == k {
      ans++
      i++
      j--
    } else if val > k {
      j--
    } else {
      i++
    }
  }
  return ans
}

File: contest/weekly/weekly-contest-218/1679_maxNumOfKSumPairs.cpp
/**
 *      file    :  1679_maxNumOfKSumPairs.cpp
 *      author  :  techiedheeraj
 *      created :  2020 Dec 06 17:37:54
 *      lastMod :  Sun Dec  6 17:37:54 2020
 **/

class Solution {
public:
    int maxOperations(vector<int>& nums, int k) {
      unordered_map<int, int> hm;

      for (auto &i: nums)
        hm[i]++;
      int c =0;

      for (int i = 0; i < nums.size(); ++i) {
        if (hm[nums[i]] < 1)
          continue;
        hm[nums[i]]--;
        if (hm.find(k-nums[i]) != hm.end() && hm[k-nums[i]]>=1) {
          c++;
          hm[k-nums[i]]--;
        }
      }
      return c;
    }
};
